engine/src/main/java/com/datatorrent/stram/codec/LogicalPlanSerializer.java

```diff
@@ -116,91 +165,144 @@ public boolean useForType(JavaType t)
+    for (StreamMeta sMeta : moduleMeta.getDag().getAllStreams()) {
+      if (sMeta.getModuleName() == null) {
+        String fullName = moduleMeta.getFullName() + LogicalPlan.MODULE_NAMESPACE_SEPARATOR + sMeta.getName();
+        StreamMeta meta = plan.getStream(fullName);
+        streams.add(meta);
       }
     }
-    Collection<StreamMeta> allStreams = dag.getAllStreams();
+    return streams;
+  }
 
-    for (StreamMeta streamMeta : allStreams) {
-      HashMap<String, Object> streamDetailMap = new HashMap<String, Object>();
-      String streamName = streamMeta.getName();
-      streamMap.add(streamDetailMap);
-      String sourcePortName = streamMeta.getSource().getPortName();
-      OperatorMeta operatorMeta = streamMeta.getSource().getOperatorMeta();
-      HashMap<String, Object> sourcePortDetailMap = new HashMap<String, Object>();
-      sourcePortDetailMap.put("operatorName", operatorMeta.getName());
-      sourcePortDetailMap.put("portName", sourcePortName);
-      streamDetailMap.put("name", streamName);
-      streamDetailMap.put("source", sourcePortDetailMap);
-      List<InputPortMeta> sinks = streamMeta.getSinks();
-      ArrayList<HashMap<String, Object>> sinkPortList = new ArrayList<HashMap<String, Object>>();
-      for (InputPortMeta sinkPort : sinks) {
-        HashMap<String, Object> sinkPortDetailMap = new HashMap<String, Object>();
-        sinkPortDetailMap.put("operatorName", sinkPort.getOperatorWrapper().getName());
-        sinkPortDetailMap.put("portName", sinkPort.getPortName());
-        sinkPortList.add(sinkPortDetailMap);
+  private static List<Map<String, Object>> getLogicalModulesInfo(LogicalPlan dag, boolean flatten)
+  {
+    List<Map<String, Object>> modulesArray = new ArrayList<Map<String, Object>>();
+    for (ModuleMeta moduleMeta : dag.getAllModules()) {
+      modulesArray.add(fillLogicalModuleDetails(dag, moduleMeta, flatten));
+    }
+    return modulesArray;
+  }
+
+  private static Map<String, Object> fillLogicalModuleDetails(LogicalPlan dag, ModuleMeta moduleMeta, boolean flatten)
+  {
+    Map<String, Object> moduleDetailMap = new HashMap<String, Object>();
+    ArrayList<Map<String, Object>> operatorArray = new ArrayList<Map<String, Object>>();
+    ArrayList<Map<String, Object>> streamArray = new ArrayList<Map<String, Object>>();
+    moduleDetailMap.put("name", moduleMeta.getName());
+    moduleDetailMap.put("className", moduleMeta.getModule().getClass().getName());
+    if (flatten) {
+      moduleDetailMap.put("operators", operatorArray);
+      for (OperatorMeta operatorMeta : getTopLevelOperators(dag, moduleMeta)) {
+        operatorArray.add(getLogicalOperatorDetails(operatorMeta));
       }
-      streamDetailMap.put("sinks", sinkPortList);
-      if (streamMeta.getLocality() != null) {
-        streamDetailMap.put("locality", streamMeta.getLocality().name());
+      moduleDetailMap.put("streams", streamArray);
+      for (StreamMeta streamMeta : getTopLevelStreams(dag, moduleMeta)) {
```

Comment:
Why iterate over all streams again? If this pattern is duplicated in more places, please fix there as well.